plugins/producer_plugin/test/test_block_timing_util.cpp

```diff
@@ -34,73 +35,219 @@ BOOST_AUTO_TEST_CASE(test_calculate_block_deadline) {
                         ninth_block_time.to_time_point() - fc::milliseconds(300));
    }
 }
 
+BOOST_AUTO_TEST_CASE(test_calculate_producer_wake_up_time) {
+   using namespace eosio;
+   using namespace eosio::chain;
+   using namespace eosio::chain::literals;
+   using namespace eosio::block_timing_util;
+
+   producer_watermarks empty_watermarks;
+   // use full cpu effort for most of these tests since calculate_producing_block_deadline is tested above
+   constexpr uint32_t full_cpu_effort = eosio::chain::config::block_interval_us;
+
+   { // no producers
+      BOOST_CHECK_EQUAL(calculate_producer_wake_up_time(full_cpu_effort, 2, chain::block_timestamp_type{}, {}, {}, empty_watermarks), std::optional<fc::time_point>{});
+   }
+   { // producers not in active_schedule
+      std::set<chain::account_name>          producers{"p1"_n, "p2"_n};
+      std::vector<chain::producer_authority> active_schedule{{"active1"_n}, {"active2"_n}};
+      BOOST_CHECK_EQUAL(calculate_producer_wake_up_time(full_cpu_effort, 2, chain::block_timestamp_type{}, producers, active_schedule, empty_watermarks), std::optional<fc::time_point>{});
+   }
+   { // Only one producer in active_schedule, we should produce every block
+      std::set<chain::account_name>          producers{"p1"_n, "p2"_n};
+      std::vector<chain::producer_authority> active_schedule{{"p1"_n}};
+      const uint32_t prod_round_1st_block_slot = 100 * active_schedule.size() * eosio::chain::config::producer_repetitions - 1;
+      for (uint32_t i = 0; i < static_cast<uint32_t>(config::producer_repetitions * active_schedule.size() * 3); ++i) { // 3 rounds to test boundaries
+         block_timestamp_type block_timestamp(prod_round_1st_block_slot + i);
+         auto block_time = block_timestamp.to_time_point();
+         BOOST_CHECK_EQUAL(calculate_producer_wake_up_time(full_cpu_effort, 2, block_timestamp, producers, active_schedule, empty_watermarks), block_time);
+      }
+   }
+   { // We have all producers in active_schedule configured, we should produce every block
+      std::set<chain::account_name>          producers{"p1"_n, "p2"_n, "p3"_n};
+      std::vector<chain::producer_authority> active_schedule{{"p1"_n}, {"p2"_n}};
+      const uint32_t prod_round_1st_block_slot = 100 * active_schedule.size() * eosio::chain::config::producer_repetitions - 1;
+      for (uint32_t i = 0; i < static_cast<uint32_t>(config::producer_repetitions * active_schedule.size() * 3); ++i) { // 3 rounds to test boundaries
+         block_timestamp_type block_timestamp(prod_round_1st_block_slot + i);
+         auto block_time = block_timestamp.to_time_point();
+         BOOST_CHECK_EQUAL(calculate_producer_wake_up_time(full_cpu_effort, 2, block_timestamp, producers, active_schedule, empty_watermarks), block_time);
+      }
+   }
+   { // We have all producers in active_schedule in active_schedule of 21 (plus a couple of extra producers configured), we should produce every block
```

Comment:
Very minor: change `in active_schedule in active_schedule` to `in active_schedule `